File: golang/src/p12/main.go
package main

import (
	"fmt"
	"math"
)

func main() {
	triangleNum := 1
	for i := 2; numFactors(triangleNum) <= 500; i++ {
		triangleNum += i
	}

	fmt.Println(triangleNum)
}

func numFactors(n int) int {
	sqrtN := int(math.Sqrt(float64(n)))
	result := 0
	for i := 1; i <= sqrtN; i++ {
		if n%i == 0 {
			if n/i == i {
				result += 1
			} else {
				result += 2
			}
		}
	}
	return result
}

File: cpp/src/p12/main.cc
#include <iostream>
#include <math.h>

int numFactors(int n)
{
  int result = 0;
  const int stop = sqrt(n);
  for (int i = 1; i <= stop; ++i) {
    if (n%i == 0) {
      result += (n/i == i ? 1 : 2);
    }
  }
  return result;
}

int main() {
  int triangleNum = 1;
  for (int i = 2; numFactors(triangleNum) <= 500; ++i) {
    triangleNum += i;
  }
  std::cout << triangleNum << "\n";
}
